Bluetooth: Controller: llcp: CTE request: Add handling and tests for LL_UNKNOWN_RSP

ppryga-nordic opened the pull request:
There was missing handling of LL_UNKNOWN_RSP in CTE request control
procedure. In case there is a pending CTE request and peer responses
with LL_UNKNOWN_RSP then Host should be notified with HCI_LE_CTE_-
Request_Failed event. The pending CTE request procedure should be
completed.

The CTE request procedure should be enabled only if a peer supports
CTE response feature. That information can be obtained by feature
exchange procedure. If there were no feature exchange then CTE
request feature may be disabled if a peer responses with LL_UNKNOWN_RSP
for a CTE request.

The implementation of ll_df_set_conn_cte_req_enable was checking if
CTE response feature is supported only when there was feature exchange.
There was missing possibility to stop CTE request if a peer responded
with LL_UNKNOWN_RSP for an earlier CTE request.

The commit changes the implementation of ll_df_set_conn_cte_req_enable.
The CTE response feature check is moved to ull_cp_cte_req function,
because it belongs more to control procedure than to function that
handles Host request to start the procedure.
Second change is related with use of conn->llcp.fex.features_used.
It stores information about features supported by peer. It does
not depend on execution of the feature exchange control procedure.

By the way, there were removed else statement in ll_df_set_conn_cte_-
req_enable because it was not needed.

Add test that verify correct behavior of the CTE request procedure
in case of reception of LL_UNKNOWN_RSP PDU from peer device.

Fixes not working CTE request control procedure unit tests.
Adds small refactoring related changes to CTE request control procedure unit tests.

There was missing code responsible for handling of unexpected response
for CTE request. The commit adds code that will terminate connection
in case a peer device responses with unexpected control PDU that is not
a remote procedure request.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/45708.

ppryga-nordic committed:
Bluetooth: Controller: llcp: add missing handling LL_UNKNOWN in CTE llcp

There was missing handling of LL_UNKNOWN_RSP in CTE request control
procedure.In case there is a pending CTE request and peer responses
with LL_UNKNOWN_RSP then Host should be notified with HCI_LE_CTE_-
Request_Failed event. The pending CTE request procedure should be
completed.
Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: df: Fix CTE reqest proc enable if feat supported

The CTE request procedure should be enabled only if a peer supports
CTE response feature. That information can be obtained by feature
exchange procedure. If there were no feature exchange then CTE
request feature may be disabled if a peer responses with LL_UNKNOWN_RSP
for a CTE request.

The implementation of ll_df_set_conn_cte_req_enable was checking if
CTE response feature is supported only when there was feature exchange.
There was missing possibility to stop CTE request if a peer responded
with LL_UNKNOWN_RSP for an earlier CTE request.

The commit changes the implementation of ll_df_set_conn_cte_req_enable.
The CTE response feature check is moved to ull_cp_cte_req function,
because it belongs more to control procedure than to function that
handles Host request to start the procedure.
Second change is related with use of conn->llcp.fex.features_used.
It stores information about features supported by peer. It does
not depend on execution of the feature exchange control procedure.

By the way, there were removed else statement in ll_df_set_conn_cte_-
req_enable because it was not needed.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: cte_req: Add tests for handling LL_UNKNOWN_RSP

Add test that verify correct behavior of the CTE request procedure
in case of reception of LL_UNKNOWN_RSP PDU from peer device.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: cte_req: Enable not used test case

There was a test case that was never executed. It was not
added to a test suite in test_main. The commit adds the
test case.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: Add missing NTF release

There were couple of test cases where Host notification object
was not released. The commit adds missing release calls.

Also commnets related with check if RX queue is empty were
changed to describe what is verified then.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: Add handling of unexpe CTRL PDU by CTE req

There was missing code responsible for handling of unexpected response
for CTE request. The commit adds code that will terminate connection
in case a peer device reposnes with unexpected control PDU that is not
a remote procedure reques.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: llcp: cte_req: remove invalid rsp test

There is a test that verifies a behavior of CTE request control
procedure if receives LL_UNKNOWN_RSP. The expected behavior
is to terminate connection if that is response handling is
not implemented by a command. Other unexpected responses are
treated as new remote control procedures.

The CTE request has implemented handling of LL_UNKNOWN_RSP,
hence this particular test is no longer valid.

There is still open question how to handle other unexpected
reposne PDUs while waiting for LL_CTE_RSP. That is not defined
by BT 5.3. Core specification and not decided by community.

The test is removed to because it fails and blocks CI.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Added commit that removes a test that was no longer valid one. More information in commit message.

ppryga commented:
@thoh-ot @cvinayak @kruithofa @erbr-ot could you take a look on the PR?

ppryga commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -722,6 +722,97 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	event_done(&conn);
+
+	/* Receive notification of sampled CTE response */
+	ut_rx_pdu(LL_UNKNOWN_RSP, &ntf, &remote_reject_ext_ind);
```
Thank you @thoh-ot! Good catch. It should be an address to `unknown_rsp` instance.

ppryga-nordic committed:
tests: Bluetooth: df: Fix wrong max antenna switch pattern len config

Due to change in Kconfig CONFIG_BT_CTLR_DF_MAX_ANT_SW_PATTERN_LEN
the Kconfigs in tests were wrong and tests were not building.
The commit sets the max antenna switch pattern length to actual
maximum value.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Add missing peer supported feature information

There was missing information about supported CTE response feature
by peer device. That caused a test_set_conn_cte_req_enable to fail.
The test expected that peer device supprots CTE response.

The missing information is added in a common code responsible for
connection object preparation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Add missing conn handle assignment

There was missing assignment to a global variable that stores
connection handle. The value was always the same, hence it test
were not failing. The value was passed over from setup state
done for other tests. This change makes sure there is always
correct handle stored in global variable.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
I have to rebase and fix merge conflicts.
Also I've spot that there were issues with unit tests, hence I've added 3 new commits that address:
- wrong value assigned to `CONFIG_BT_CTLR_DF_MAX_ANT_SW_PATTERN_LEN`
- missing CTE response feature in features supported by peer device
- missing connection handle assignment to global connection handle variable

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -971,6 +971,13 @@ uint8_t ull_cp_cte_req(struct ll_conn *conn, uint8_t min_cte_len, uint8_t cte_ty
+	/* If controller gained,by any means, awareness of features supported by peer device
+	 * then check whether required features are enabled.
+	 */
+	if (!(conn->llcp.fex.features_used & BIT64(BT_LE_FEAT_BIT_CONN_CTE_RESP))) {
```
Good catch; `llcp.fex.features_used` is initialized to the features of the local controller (`LL_FEAT`) and without checking `llcp.fex.valid` you cannot assume anything about the peer controller.

thoh-ot replied:
Maybe also look into _ull_llcp_features.h_ and add any missing feature helper functions for CTE

ppryga commented:
Fixed issue found by @cvinayak related with support for CTE RSP by peer device.
